src/libraries/System.Private.CoreLib/src/System/Reflection/ConstructorInvoker.cs

```diff
@@ -2,25 +2,94 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.Reflection
 {
     internal sealed partial class ConstructorInvoker
     {
-        private readonly RuntimeConstructorInfo _constructorInfo;
-        public InvocationFlags _invocationFlags;
+        private readonly RuntimeConstructorInfo _method;
+
+#if !MONO // Temporary until Mono is updated.
+        private bool _invoked;
+        private bool _strategyDetermined;
+        private InvokerEmitUtil.InvokeFunc? _emitInvoke;
+#endif
 
         public ConstructorInvoker(RuntimeConstructorInfo constructorInfo)
         {
-            _constructorInfo = constructorInfo;
+            _method = constructorInfo;
+
+#if USE_NATIVE_INVOKE
+            // Always use the native invoke; useful for testing.
+            _strategyDetermined = true;
+#elif USE_EMIT_INVOKE
+            // Always use emit invoke (if IsDynamicCodeCompiled == true); useful for testing.
+            _invoked = true;
+#endif
         }
 
+#if MONO // Temporary until Mono is updated.
+        [DebuggerStepThrough]
+        [DebuggerHidden]
+        public unsafe object? InvokeUnsafe(object? obj, Span<object?> args, BindingFlags invokeAttr)
+        {
+            return InvokeNonEmitUnsafe(obj, args, invokeAttr);
+        }
+#else
         [DebuggerStepThrough]
         [DebuggerHidden]
-        public unsafe object? InvokeUnsafe(object? obj, IntPtr* args, Span<object?> argsForTemporaryMonoSupport, BindingFlags invokeAttr)
+        public unsafe object? InvokeUnsafe(object? obj, IntPtr* args, BindingFlags invokeAttr)
         {
-            // Todo: add strategy for calling IL Emit-based version
-            return _constructorInfo.InvokeNonEmitUnsafe(obj, args, argsForTemporaryMonoSupport, invokeAttr);
+            if (!_strategyDetermined)
+            {
+                if (!_invoked)
+                {
+                    // The first time, ignoring race conditions, use the slow path.
+                    _invoked = true;
+                }
+                else
+                {
+                    if (RuntimeFeature.IsDynamicCodeCompiled)
+                    {
+                        _emitInvoke = InvokerEmitUtil.CreateInvokeDelegate(_method);
+                    }
+                    _strategyDetermined = true;
+                }
+            }
+
+            object? ret;
+            if ((invokeAttr & BindingFlags.DoNotWrapExceptions) == 0)
+            {
+                try
+                {
+                    // For the rarely used and broken scenario of calling the constructor directly through MethodBase.Invoke()
```

Comment:
I agree that this is rarely used, but is it really broken?

E.g. if you were to built C# interpreter using reflection, you need ability to call base-class constructor on an already allocated object.

Reply:
Thanks. Updated